### PR TITLE
EZP-28739: Add continue on errors option and display corrupted content IDs in the search indexer

### PR DESCRIPTION
**Pull request by kmadejski**
…t IDs in the search indexer

JIRA issue: https://jira.ez.no/browse/EZP-28739

# Description
As mentioned in JIRA issue, this improvement allows a user to see exactly which content IDs were unindexable. Moreover, this PR adds `--continue-on-error` option to the indexer.

- [x] Change `ezsystems/ezplatform-solr-search-engine` accordingly: https://github.com/ezsystems/ezplatform-solr-search-engine/pull/122

Before:
screen shot 2018-06-11 at 14 22 15

After (with `--continue-on-error`):
screen shot 2018-06-11 at 14 19 59

After (without `--continue-on-error`):
screen shot 2018-06-11 at 14 31 15

### CODE REVIEW

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 1]
```diff
@@ -233,12 +244,16 @@ protected function indexIncrementally(InputInterface $input, OutputInterface $ou
...
-                $this->searchIndexer->updateSearchIndex($contentIds, $commit);
+                $unindexableContentIds += $this->searchIndexer->updateSearchIndex($contentIds, $commit, $continueOnError);
                 $progress->advance(1);
             }
         }
 
         $progress->finish();
+
+        if ($unindexableContentIds) {
+            $this->printUnindexableContentIds($output, $unindexableContentIds);
```
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: missing return for error code, also need to decide in what conditions parallel processes indexing should fail or not, as it afaik will otherwise silently index all and not make user aware there where errors.

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 2]
```diff
@@ -433,4 +448,17 @@ private function getNumberOfCPUCores()
...
+    /**
+     * @param OutputInterface $output
+     * @param array $unindexableContentIds
+     */
+    private function printUnindexableContentIds(OutputInterface $output, $unindexableContentIds)
+    {
+        $output->writeln('');
+        $output->writeln(sprintf(
+            '<error>Indexing failed on some content items, try running command with "--iteration-count=1 --content-ids=<ids>": %s</error>',
+            implode(', ', $unindexableContentIds)
```
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: If we skip the space after `,` it will be possible to copy paste the output to --content-ids argument.

File: eZ/Publish/Core/Search/Legacy/Content/Indexer.php [hunk 3]
```diff
@@ -46,8 +48,16 @@ public function updateSearchIndex(array $contentIds, $commit)
                 }
             } catch (NotFoundException $e) {
                 $this->searchHandler->deleteContent($contentId);
+            } catch (InvalidArgumentValue $argumentException) {
+                $this->logger->warning($argumentException->getMessage());
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Indexer.php]: warning in case of `continueOnError`, error if not.

File: eZ/Publish/Core/FieldType/RichText/SearchField.php [hunk 4]
```diff
@@ -27,11 +28,15 @@ class SearchField implements Indexable
...
      * @return \eZ\Publish\SPI\Search\Field[]
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentValue if XML is corrupted
      */
     public function getIndexData(Field $field, FieldDefinition $fieldDefinition)
     {
         $document = new DOMDocument();
-        $document->loadXML($field->value->data);
+        if ($document->loadXML($field->value->data) === false) {
+            throw new InvalidArgumentValue($field->type, 'Unable to load XML', self::class);
```
Comment (andrerom) [inline eZ/Publish/Core/FieldType/RichText/SearchField.php]: This change should probably be separate PR

**Review by andrerom — CHANGES_REQUESTED** [review]
Unsure here:
1. it seems to break parallel indexing in current form as it will catch the exception and hence indexer will just shug along and user won't see the output coming form the child process with --content-ids argument.
2. other exceptions is not handled so makes this a bit strange, I would expect it to handle all exceptions, however unsure if Indexer should do that _(if it's Indexer responsibility)_

@alongosz Maybe you have some input here?

File: eZ/Publish/SPI/FieldType/Indexable.php [hunk 5]
```diff
@@ -24,6 +24,8 @@
      * @param \eZ\Publish\SPI\Persistence\Content\Type\FieldDefinition $fieldDefinition
      *
      * @return \eZ\Publish\SPI\Search\Field[]
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentValue
```
Comment (adamwojs) [inline eZ/Publish/SPI/FieldType/Indexable.php]: How about introducing more specific exception (something like `IndexingExcetption`/`CorruptedDataException` ) with the more precise location of the error (id, version, language code)?

**Review by adamwojs — CHANGES_REQUESTED** [review]

File: eZ/Publish/SPI/FieldType/Indexable.php (hunk 5 above)
Comment (kmadejski) [reply eZ/Publish/SPI/FieldType/Indexable.php]: I'm totally okay with the idea of introduction of the new specific exception. What do you think @andrerom?

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php (hunk 1 above)
Comment (kmadejski) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: Yes, you are absolutely right. I'm going to rewrite this part.

File: eZ/Publish/SPI/FieldType/Indexable.php (hunk 5 above)
Comment (andrerom) [reply eZ/Publish/SPI/FieldType/Indexable.php]: I'm ok with it as well, so we can document it on SearchField, then it becomes much more natural to catch it in indexer.

Only question is: naming :P
To identify that, what it the intent here? given the method it is thrown from maybe something like `[Invalid]IndexDataException`

Comment (kmadejski) [reply eZ/Publish/SPI/FieldType/Indexable.php]: For me `InvalidIndexDataException` sounds great 👍 

Where should it be placed?
a) eZ\Publish\Core\Base\Exceptions
b) eZ\Publish\API\Repository\Exceptions
or should we introduce new namespace:
c) eZ\eZ\Publish\SPI\Exceptions

Comment (adamwojs) [reply eZ/Publish/SPI/FieldType/Indexable.php]: +1 for `InvalidIndexDataException` and C

**Comment by andrerom** [comment]
Side: One of the files has gotten it's permissions changed `100644 → 100755`

File: eZ/Publish/SPI/FieldType/Indexable.php (hunk 5 above)
Comment (andrerom) [reply eZ/Publish/SPI/FieldType/Indexable.php]: I would go for `eZ\Publish\SPI\FieldType\Exceptions\`

Think of SPI folders as potentially separate packages.

File: eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * @return string
+     */
+    public function getLanguageCode()
+    {
+        return $this->languageCode;
+    }
+}
```
Comment (andrerom) [inline eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException.php]: nitpick: new line

Comment (kmadejski) [reply eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException.php]: File permissions and nitpick fixed. Thanks.

**Comment by andrerom** [comment]
So the one case of exception is now pretty clean, however there are other exceptions can should be handled by a broad reaching `continue-on-error` _(as opposed to `continue-on-eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException` ;) )_

One of them is missing field type, so in case where user has not yet configured null type.
And there might be others.

**Comment by lserwatka** [comment]
@kmadejski what is the status on this one?

File: eZ/Publish/Core/Search/Common/IncrementalIndexer.php [hunk 7]
```diff
@@ -73,8 +73,9 @@
      *
      * @param int[] $contentIds
      * @param bool $commit
+     * @param bool $continueOnError
      */
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Common/IncrementalIndexer.php]: Add doc for return value

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 8]
```diff
@@ -239,7 +239,7 @@ protected function indexIncrementally(InputInterface $input, OutputInterface $ou
         $progress = new ProgressBar($output);
         $progress->start($iterations);
 
-        if ($processCount > 1) {
+        if ($processCount > 1 && !$continueOnError) {
```
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: you should probably take this into account earlier, otherwise `$processMessage` above will be wrong.

File: eZ/Publish/Core/Search/Common/IncrementalIndexer.php [hunk 9]
```diff
@@ -74,6 +74,7 @@
      * @param int[] $contentIds
      * @param bool $commit
      * @param bool $continueOnError
+     * @return array
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Common/IncrementalIndexer.php]: `@return int[] Returns id's content items we could not index as field type indexer threw a InvalidIndexDataException, details on the exception can be found in log message.`

File: eZ/Publish/Core/Search/Common/Indexer.php [hunk 10]
```diff
@@ -46,16 +47,23 @@
...
         DatabaseHandler $databaseHandler,
-        SearchHandler $searchHandler
+        SearchHandler $searchHandler,
+        IndexerErrorHandler $errorCollector = null
     ) {
         $this->logger = $logger;
         $this->persistenceHandler = $persistenceHandler;
         $this->databaseHandler = $databaseHandler;
         $this->searchHandler = $searchHandler;
+        $this->errorHandler = $errorCollector ?: new NullErrorHandler();
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Common/Indexer.php]: handler or collector? lets pick one, Collector might be the clearest word here for what it does.

wdyt @alongosz ?

File: eZ/Publish/Core/Search/Common/IndexerErrorHandler.php [hunk 11]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace eZ\Publish\Core\Search\Common;
+
+use eZ\Publish\SPI\Persistence\Content\ContentInfo;
+
+/**
+ * Interface for handling errors during indexing.
+ */
+interface IndexerErrorHandler
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Common/IndexerErrorHandler.php]: Missing hasErrors and getErrors

File: eZ/Publish/Core/Search/Common/IndexerErrorHandler/NullErrorHandler.php [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+namespace eZ\Publish\Core\Search\Common\IndexerErrorHandler;
+
+use eZ\Publish\Core\Search\Common\IndexerErrorHandler;
+use eZ\Publish\SPI\Persistence\Content\ContentInfo;
+
+class NullErrorHandler implements IndexerErrorHandler
+{
+    public function handle(ContentInfo $contentInfo, $errorMessage)
+    {
+        return false;
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Common/IndexerErrorHandler/NullErrorHandler.php]: continue = false?

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 13]
```diff
@@ -54,7 +60,11 @@ class ReindexCommand extends ContainerAwareCommand
     public function initialize(InputInterface $input, OutputInterface $output)
     {
         parent::initialize($input, $output);
-        $this->searchIndexer = $this->getContainer()->get('ezpublish.spi.search.indexer');
+
+        $searchIndexerFactory = $this->getContainer()->get('ezpublish.api.search_engine.indexer.factory');
+        $this->errorHandler = new LocalErrorHandler($input->getOption('continue-on-error'));
+        $this->searchIndexer = $searchIndexerFactory->buildSearchEngineIndexerWithErrorHandler($this->errorHandler);
```
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: Would perhaps be clearer if either:
A. get indexer like before and simply call the new setter method for setting the optional error handler
B. Maybe remove the continue-on-error handling from ErrorHandler so it can be injected by DI? If so it will remove the hard coding of handler here and will instead need continue-on-error flag to be sent to indexer
C. ..

wdyt @alongosz & @adamwojs ?

File: eZ/Publish/Core/Search/Common/Indexer.php (hunk 10 above)
Comment (alongosz) [reply eZ/Publish/Core/Search/Common/Indexer.php]: > handler or collector? lets pick one, Collector might be the clearest word here for what it does.

Well... the interface has the `handle` method, so I'd say handler. But AFAICS `handle` only collects these errors, right? In that case it should be called `collect` and the interface and implementing classes - `*Collector`

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php (hunk 13 above)
Comment (alongosz) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: Search Indexer with proper error handler (collector? :grinning: ) should be injected via DI.

Comment (alongosz) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: On second thought, since building this service depends on `continue-on-error` command option, it might be impossible to do it completely via DI. However still factories shouldn't be used directly in a such way.
So I'd rather simplify this to:
```php
$this->searchIndexer->setErrorHandler(new LocalErrorHandler($input->getOption('continue-on-error')))
````
Of course instead of `new LocalErrorHandler` you could create factory for error handler, but that's maybe an overkill, since only `LocalErrorHandler` provides strategy to continue on errors.

To be safe, before calling setErrorHandler I'd heck if `$searchIndexer instanceof IncrementalIndexer`.

Side note: could you refactor this command to get only needed services instead of entire container? When I wrote it I lacked proper knowledge :sweat_smile:

Comment (andrerom) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: This is A.

B would be to move continue-on-error logic injection to indexer or somewhere else so collector does not need to know about this, and instead just collects if called.

File: eZ/Publish/Core/Search/Common/IndexerErrorHandler/NullErrorHandler.php (hunk 12 above)
Comment (kmadejski) [reply eZ/Publish/Core/Search/Common/IndexerErrorHandler/NullErrorHandler.php]: Yes, that's true.

File: eZ/Publish/Core/Search/Common/IndexerErrorHandler.php (hunk 11 above)
Comment (kmadejski) [reply eZ/Publish/Core/Search/Common/IndexerErrorHandler.php]: Added in https://github.com/ezsystems/ezpublish-kernel/pull/2221/commits/4f8d3f0543ea1ddaae33c42a945ab2cc63c2def2.

File: eZ/Publish/Core/Search/Common/Indexer.php (hunk 10 above)
Comment (kmadejski) [reply eZ/Publish/Core/Search/Common/Indexer.php]: Renamed to `Collector` in https://github.com/ezsystems/ezpublish-kernel/pull/2221/commits/4f8d3f0543ea1ddaae33c42a945ab2cc63c2def2.

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 14]
```diff
@@ -207,7 +235,7 @@ protected function indexIncrementally(InputInterface $input, OutputInterface $ou
         }
 
         $iterations = ceil($count / $iterationCount);
-        $processes = $input->getOption('processes');
+        $processes = $continueOnError ? 1 : $input->getOption('processes');
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: If `continue-on-error` doesn't allow for `processes` setting, it should be stated in a command doc also.

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 15]
```diff
@@ -433,4 +463,18 @@ private function getNumberOfCPUCores()
 
         return $cores;
     }
+
+    /**
+     * @param OutputInterface $output
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: Nitpick: please use FQCN

File: eZ/Publish/Core/Search/Common/IndexerErrorCollector.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace eZ\Publish\Core\Search\Common;
```
Comment (alongosz) [inline eZ/Publish/Core/Search/Common/IndexerErrorCollector.php]: Nitpick: missing license/copyright.

File: eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException.php [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    protected $languageCode;
+
+    public function __construct($fieldId, $versionNo, $languageCode, Exception $previous = null)
+    {
+        $this->fieldId = $fieldId;
+        $this->versionNo = $versionNo;
+        $this->languageCode = $languageCode;
+
+        $message = sprintf('Field %d in the version %d for %s language can not be indexed', $this->fieldId, $this->versionNo, $this->languageCode);
+        parent::__construct($message, 0, $previous);
```
Comment (alongosz) [inline eZ/Publish/SPI/FieldType/Exceptions/InvalidIndexDataException.php]: Really small nitpick: we shouldn't use 0 for error code (though I saw we often do). 0 means success (for Mac also, so you can't say I'm biased ;P). Due to lack of better idea I usually use 1. This is of course pure matter of convention.

**Review by alongosz — APPROVED** [review]
I'm fine with the current approach (designed as "A").

But if you want to redesign it so:
> B. (...) move continue-on-error logic injection to indexer or somewhere else so collector does not need to know about this, and instead just collects if called.

then by all means, do it, please :)

Nitpicks:

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php (hunk 14 above)
Comment (kmadejski) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: I need to check again whether with this new approach this is still relevant. If yes, then I'll add a note to command docs.

Comment (kmadejski) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: I've tested it, and it's not going to work with parallel processing even with this new approach.
 
The information which you've asked for is currently in the command option description:
https://github.com/ezsystems/ezpublish-kernel/pull/2221/files#diff-81257ebfd63be05261b508add93d35edR137
It is enough, isn't it? Or would you like to add it also to command help, perhaps?

Comment (alongosz) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: I would also be good to update `processes` option help stating that it cannot be used with `continue-on-errors`.

Comment (kmadejski) [reply eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: Added in https://github.com/ezsystems/ezpublish-kernel/pull/2221/commits/91594d961baf39a4937c03dff479b64595de611d.

**Comment by kmadejski** [comment]
I have no idea why this test https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/392127163 didn't pass. There was only one change in command option description. All tests passed for the previous commit: https://travis-ci.org/ezsystems/ezpublish-kernel/builds/391290536. Seems to not be related to any of my changes.

**Comment by andrerom** [comment]
@kmadejski If it's the language test on REST test suite then that is my blame from http-cache, for now avoided it here until I can debug where it comes from in http-cache 0.7: https://github.com/ezsystems/ezplatform/commit/70d6fb7042c4e68b2d3fedf7db457f0aed620a04

**Review by adamwojs — COMMENTED** [review]
Could you also please add a unit tests for changes introduced in this PR ?

**Comment by kmadejski** [comment]
@adamwojs tests added in https://github.com/ezsystems/ezpublish-kernel/pull/2221/commits/20b9289284dd2be85312ad257b89f188d4f2823c.

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 18]
```diff
@@ -54,9 +61,20 @@ class ReindexCommand extends ContainerAwareCommand
     public function initialize(InputInterface $input, OutputInterface $output)
     {
         parent::initialize($input, $output);
-        $this->searchIndexer = $this->getContainer()->get('ezpublish.spi.search.indexer');
+
+        $searchIndexer = $this->getContainer()->get('ezpublish.spi.search.indexer');
+        // It's required to clone $searchIndexer, otherwise global service state will be changed due to setting errorCollector
+        $this->searchIndexer = clone $searchIndexer;
```
Comment (andrerom) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: For me this is kind of design smell, but unsure how much more it's needed to iterate on this.

One way to avoid would be if continue-on-error knowledge is moved to indexer, so all the stuff on bool return value on ErrorCollector interface and such can be removed simplifying it to truly be only about collecting errors.

And instead end up with `->updateSearchIndex($contentIds, $commit, $continueOnError)` so we don't need to clone anything as it will more cleanly be a normal call argument like it should be.

But it's a new argument, so A. either needs to be introduced as commented out `/*bool $continueOnError= false*/` and with php doc stating it will be added to interface in next major _(with inline deprecation notice or something we can find when next major comes)_. Symfony sometimes does this on addition/removal of arguments, so for me this is fine, impl can always add arguments interface does not have as long as it has default value.

Or B. we need to add yet another interface and deprecate the old one.

I think I prefer A as otherwise we keep on adding new interfaces all the time, and it is a bit softer dependency for dealing with kernel vs solr bundle.

**Review by andrerom — CHANGES_REQUESTED** [review]

**Review by konradoboza — APPROVED** [review]
Option's name is `continue-on-error` so the titles of PRs and ticket are a bit confusing. :)

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 19]
```diff
@@ -69,11 +76,22 @@ class ReindexCommand extends ContainerAwareCommand
     public function initialize(InputInterface $input, OutputInterface $output)
     {
         parent::initialize($input, $output);
-        $this->searchIndexer = $this->getContainer()->get('ezpublish.spi.search.indexer');
+
+        $searchIndexer = $this->getContainer()->get('ezpublish.spi.search.indexer');
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: This is obsolete, unless you have solution ready to apply when merging into master, I won't accept it right now. It needs to be injected.

File: eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php [hunk 20]
```diff
@@ -69,11 +76,22 @@ class ReindexCommand extends ContainerAwareCommand
...
+        // It's required to clone $searchIndexer, otherwise global service state will be changed due to setting errorCollector
+        $this->searchIndexer = clone $searchIndexer;
+        if ($this->searchIndexer instanceof IncrementalIndexer) {
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php]: This violates open-closed principle and moreover should be handled by DIC.

**Review by alongosz — COMMENTED** [review]
Another round of review due to alternative suggested via #2666

**Comment by kmadejski** [comment]
@alongosz thanks for taking a glance on this again! 

Since this PR is quite old and as far as I remember we couldn't really agree on all concepts proposed here and there, I think it would be great if we can schedule some time to sit to the table together (maybe with @andrerom and @adamwojs if possible) and discuss what and how we want to achieve here. I would gladly continue work on this because I still see it important and useful for the customers.

**Comment by andrerom** [comment]
@kmadejski Lets see if we can do that tomorrow afternoon, ideally with feedback on the other PR to clarify if he had issues with same issue or not. If so I prefer this cleaner approach, but we can still consider catching everything if we want, or we combine them making documented exception handled by default, and all other exceptions only ignored when using `--continue-on-error` 🤷‍♂

**Comment by adamwojs** [comment]
Closing PR as obsolete. eZ Platform 2.5 has reached EOM, so please reopen PR in ezsystems/ezplatform-kernel (bug fixes) or ibexa/core (features/improvements) if issue is still valid for v3.3 / v4.x and you are willing to work on it.
